common/djangoapps/entitlements/models.py

```diff
@@ -336,45 +347,59 @@ def get_fullfillable_entitlement_for_user(cls, user, course_run_key):
-                    ).exclude(
-                        expired_at__isnull=False,
-                        enrollment_course_run__isnull=False
-                    ).latest('created')
-                except CourseEntitlement.DoesNotExist:
-                    return None
-
-                # Verify that the enrollment can be enrolled in
-                if not is_course_run_entitlement_fullfillable(
-                        course_run_key=course_run_key,
-                        entitlement=course_entitlement
-                ):
-                    return None
-                return course_entitlement
+                for entitlement in entitlements:
+                    if course_uuid == entitlement.course_uuid:
+                        # Verify that the enrollment can be enrolled in
+                        if is_course_run_entitlement_fullfillable(
+                                course_run_key=course_run_key,
+                                entitlement=entitlement
+                        ):
+                            return entitlement
         return None
 
-    @classmethod
-    def check_for_existing_entitlement_and_enroll(cls, user, course_run_key):
-        entitlement = cls.get_fullfillable_entitlement_for_user(user, course_run_key)
-        if entitlement:
-            cls.enroll_user_in_course(user, entitlement, course_run_key)
-
     @classmethod
     @transaction.atomic
     def enroll_user_in_course(cls, user, entitlement, course_run_key):
-        enrollment = CourseEnrollment.enroll(
-            user=user,
-            course_key=course_run_key,
-            mode=entitlement.mode
-        )
+        """
+        Enrolls the user in the Course Run and updates the entitlement with the new Enrollment.
+
+        Returns:
+            bool: True if no errors are raised during the enroll.
+        """
+        try:
+            enrollment = CourseEnrollment.enroll(
+                user=user,
+                course_key=course_run_key,
+                mode=entitlement.mode
+            )
+        except CourseEnrollmentException:
+            log.exception('Login for Course Entitlement {uuid} failed'.format(uuid=entitlement.uuid))
+            return False
+
         entitlement.set_enrollment(enrollment)
+        return True
+
+    @classmethod
+    def check_for_existing_entitlement_and_enroll(cls, user, course_run_key):
+        """
+        Looks at the User's existing entitlements to see if the user already has a Course Entitlement for the
+        course run provided in the course_key.  If the user does have an Entitlement with no run set, the User is
+        enrolled in the mode set in the Entitlement.
+
+        Arguments:
+            user (User): The user that we are inspecting the entitlements for.
+            course_run_key (CourseKey): The course run Key.
+        Returns:
+            bool: True if the User was enrolled in the Course and the Entitlement was used, False otherwise.
```

Comment:
For clarity I would describe this truth statement as "True if the user had an eligible course entitlement to which an enrollment in the given course run was applied"